src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewTemporaryFileCommand.cs

```
@@ -2,44 +2,103 @@
+                        new ArgumentException(
+                                StringUtil.Format(UtilityCommonStrings.InvalidCharacterInParameter,
+                                    nameof(Extension), Extension)),
+                        "NewTemporaryInvalidArgument",
+                        ErrorCategory.InvalidArgument,
+                        Extension));
+                return;
+            }
+
+            string temporaryFilePath = null;
+            string tempPath = Path.GetTempPath();
+
             if (ShouldProcess(tempPath))
             {
-                try
+                int attempts = 0;
+                bool creationOfFileSuccessful = false;
+
+                // In case the random temporary file already exists, retry 
+                while (attempts++ < 10 && !creationOfFileSuccessful)
                 {
-                    filePath = Path.GetTempFileName();
+                    try
+                    {
+                        temporaryFilePath = Path.GetTempFileName(); // this already creates the temporary file
+                        if (Extension != defaultExtension)
+                        {
+                            // rename file
+                            var temporaryFileWithCustomExtension = Path.ChangeExtension(temporaryFilePath, Extension);
+                            File.Move(temporaryFilePath, temporaryFileWithCustomExtension);
+                            temporaryFilePath = temporaryFileWithCustomExtension;
+                        }
+                        creationOfFileSuccessful = true;
+                        WriteDebug($"Created temporary file {temporaryFilePath}.");
+                    }
+                    catch (IOException) // file already exists -> retry
+                    {
+                        attempts++;
+                        if (temporaryFilePath != null)
+                        {
+                            File.Delete(temporaryFilePath);
+                        }
```

What if this throws? We won't retry.

---

Should we then simply try File.Delete with an empty catch block instead?

---

If the retry count is 10, that seems fine. If it was a huge number, then I'd say probably not.  A short sleep in the catch might not hurt either, like 10ms or something.

---

Would it make sense to write a warning if we're not able to delete it?

---

At least a warning seems reasonable. But I am slightly more leading to actually throwing a (terminating) error because if deleting the temporary file that just got created failed then I suspect that in most cases there is rather a fundamental problem with permissions. WDYT?